aten/src/ATen/Error.h

```diff
@@ -5,14 +5,164 @@
+
+  int status = -1;
+  // This function will demangle the mangled function name into a more human
+  // readable format, e.g. _Z1gv -> g().
+  // More information:
+  // https://github.com/gcc-mirror/gcc/blob/master/libstdc%2B%2B-v3/libsupc%2B%2B/cxxabi.h
+  char* demangled_function_name = abi::__cxa_demangle(
+      mangled_function_name.c_str(),
+      /*__output_buffer=*/nullptr,
+      /*__length=*/0,
+      &status);
+
+  // Demangling may fail, for example when the name does not follow the
+  // standard C++ (Itanium ABI) mangling scheme. This is the case for `main` or
+  // `clone` for example, so the mangled name is a fine default.
+  if (status == 0) {
+    frame.function_name = demangled_function_name;
+  } else {
+    frame.function_name = mangled_function_name;
+  }
+
+  // `__cxa_demangle` returns a malloc'd string that we have to free ourselves.
+  free(demangled_function_name);
+
+  return frame;
+}
+
+inline std::string get_backtrace(
+    size_t frames_to_skip = 0,
+    size_t maximum_number_of_frames = 64) {
+  // We always skip this frame (backtrace).
+  frames_to_skip += 1;
+
+  std::vector<void*> callstack(
+      frames_to_skip + maximum_number_of_frames, nullptr);
+  // backtrace() gives us a list of return addresses in the current call stack.
+  // NOTE: As per man (3) backtrace it can never fail
+  // (http://man7.org/linux/man-pages/man3/backtrace.3.html).
+  auto number_of_frames = ::backtrace(callstack.data(), callstack.size());
+
+  // Skip as many frames as requested. This is not efficient, but the sizes here
+  // are small and it makes the code nicer and safer.
+  for (; frames_to_skip > 0 && number_of_frames > 0;
+       --frames_to_skip, --number_of_frames) {
+    callstack.erase(callstack.begin());
+  }
+
+  // `number_of_frames` is strictly less than the current capacity of
+  // `callstack`, so this is just a pointer subtraction and makes the subsequent
+  // code safer.
+  callstack.resize(number_of_frames);
+
+  // `backtrace_symbols` takes the return addresses obtained from `backtrace()`
+  // and fetches string representations of each stack. Unfortunately it doesn't
+  // return a struct of individual pieces of information but a concatenated
+  // string, so we'll have to parse the string after.
+  char** raw_symbols = ::backtrace_symbols(callstack.data(), callstack.size());
+  std::vector<std::string> symbols(raw_symbols, raw_symbols + callstack.size());
+
+  // The array returned by `backtrace_symbols` is malloc'd and must be manually
+  // freed, but not the strings inside the array.
+  free(raw_symbols);
+
+  // The backtrace string goes into here.
+  std::ostringstream stream;
+
+  for (size_t frame_number = 0; frame_number < callstack.size();
+       ++frame_number) {
+    const auto frame = parse_frame_information(symbols[frame_number]);
+
+    // These are usually uninteresting frames below `main` or `clone` with
+    // limited debugging information.
+    if (frame.function_name.empty()) {
```

Comment:
Can we print sth like `<5 frames without debug info>` at the end? In case there would be any errors in the parsing code, that would at least give you a hint that it's this conditioned that get triggered, and it's not like the stack is all messed up

Reply:
Ok